gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerClusterConfigurationMonitor.java

```diff
@@ -0,0 +1,940 @@
+    } finally {
+      serviceConfigurationsLock.writeLock().unlock();
+    }
+  }
+
+  private void init() {
+    // Load any persisted discovery configuration data
+    loadDiscoveryConfiguration();
+
+    // Load any persisted cluster service configuration data
+    loadServiceConfiguration();
+  }
+
+  /**
+   * Get a DiscoveryApiClient for the ClouderaManager instance described by the specified discovery configuration.
+   *
+   * @param discoveryConfig The discovery configuration for interacting with a ClouderaManager instance.
+   */
+  private DiscoveryApiClient getApiClient(final ServiceDiscoveryConfig discoveryConfig) {
+    return clients.computeIfAbsent(discoveryConfig.getAddress(),
+                                   c -> new DiscoveryApiClient(discoveryConfig, aliasService));
+  }
+
+  /**
+   * Load any previously-persisted service discovery configurations.
+   */
+  private void loadDiscoveryConfiguration() {
+    File persistenceDir = getPersistenceDir();
+    if (persistenceDir != null) {
+      Collection<File> persistedConfigs = FileUtils.listFiles(persistenceDir, new String[]{"conf"}, false);
+      for (File persisted : persistedConfigs) {
+        Properties props = new Properties();
+        try (InputStream in = Files.newInputStream(persisted.toPath())) {
+          props.load(in);
+
+          addDiscoveryConfig(props.getProperty(PROP_CLUSTER_NAME), new ServiceDiscoveryConfig() {
+            @Override
+            public String getAddress() {
+              return props.getProperty(PROP_CLUSTER_SOURCE);
+            }
+
+            @Override
+            public String getUser() {
+              return props.getProperty(PROP_CLUSTER_USER);
+            }
+
+            @Override
+            public String getPasswordAlias() {
+              return props.getProperty(PROP_CLUSTER_ALIAS);
+            }
+          });
+        } catch (IOException e) {
+          log.failedToLoadClusterMonitorServiceDiscoveryConfig(getType(), e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Load any previously-persisted cluster service configuration data records, so the monitor can check
+   * previously-deployed topologies against the current cluster configuration, even across gateway restarts.
+   */
+  private void loadServiceConfiguration() {
+    File persistenceDir = getPersistenceDir();
+    if (persistenceDir != null) {
+      Collection<File> persistedConfigs = FileUtils.listFiles(persistenceDir, new String[]{"ver"}, false);
+      for (File persisted : persistedConfigs) {
+        try (ObjectInputStream in = new ObjectInputStream(Files.newInputStream(persisted.toPath()))) {
+          String clusterName      = (String) in.readObject();
+          String discoveryAddress = (String) in.readObject();
+          Map<String, ServiceConfigurationProperties> configs =
+              (Map<String, ServiceConfigurationProperties>) in.readObject();
```

Comment:
Does this need to be Objects? Can't this be a JSON file or something else?

Reply:
Good point. I suppose I was rushing a bit to get this finished. It would be better to use an alternative serialization format.

Reply:
Added JSON serialization format instead